File: app/src/main/java/pt/simdea/guestlist/MainActivity.java
package pt.simdea.guestlist;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.objectbox.Box;
import io.objectbox.BoxStore;
import io.objectbox.query.Query;
import pt.simdea.gmlrva.lib.GenericMultipleLayoutAdapter;

public class MainActivity extends AppCompatActivity implements RecyclerItemClickListener.OnItemClickListener {

    @BindView(R.id.listacompras)
    protected RecyclerView mRecyclerView;
    @BindView(R.id.search_view)
    protected MaterialSearchView mSearchView;
    private Box<Item> itemsBox;
    private Query<Item> itemsQuery;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        BoxStore boxStore = ((App) getApplication()).getBoxStore();
        itemsBox = boxStore.boxFor(Item.class);

        itemsQuery = itemsBox.query().order(Item_.id).build();

        mSearchView.setVoiceSearch(false);
        mSearchView.setOnQueryTextListener(new MaterialSearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                addElement(query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                //Do some magic
                return false;
            }
        });

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setItemViewCacheSize(30);
        mRecyclerView.setDrawingCacheEnabled(true);
        mRecyclerView.setDrawingCacheQuality(View.DRAWING_CACHE_QUALITY_HIGH);
        mRecyclerView.setNestedScrollingEnabled(false);
        mRecyclerView.setLayoutManager(linearLayoutManager);
        mRecyclerView.setAdapter(new GenericMultipleLayoutAdapter(itemsQuery.find(), this, false));
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
                final int position = viewHolder.getAdapterPosition(); //get position which is swipe

                if (mRecyclerView.getAdapter().getItemCount() > 0) {
                    GenericMultipleLayoutAdapter adapter = (GenericMultipleLayoutAdapter) mRecyclerView.getAdapter();
                    Item item = (Item) adapter.get(position);
                    if (item != null) {
                        adapter.remove(item);
                        itemsBox.remove(item);
                    }
                }
            }
        });
        itemTouchHelper.attachToRecyclerView(mRecyclerView); //set swipe to recycler view
        mRecyclerView.addOnItemTouchListener(new RecyclerItemClickListener(this, this));
    }

    public void addElement(@NonNull String name) {
        // save the new Item to the database
        if (!name.equals("")) {
            Item item = new Item(name, 1);
            GenericMultipleLayoutAdapter adapter = (GenericMultipleLayoutAdapter) mRecyclerView.getAdapter();
            itemsBox.put(item);
            adapter.add(item);
            adapter.notifyDataSetChanged();
        } else
            Toast.makeText(this, "Item vazio", Toast.LENGTH_LONG).show();
    }

    @Override
    public void onItemClick(View view, int position) {
        GenericMultipleLayoutAdapter adapter = (GenericMultipleLayoutAdapter) mRecyclerView.getAdapter();
        Item item = (Item) adapter.get(position);
        item.addCount();
        itemsBox.put(item);
        adapter.notifyDataSetChanged();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);

        MenuItem item = menu.findItem(R.id.add);
        mSearchView.setMenuItem(item);

        return true;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onBackPressed() {
        if (mSearchView.isSearchOpen()) {
            mSearchView.closeSearch();
        } else {
            super.onBackPressed();
        }
    }
}

File: app/src/main/kotlin/pt/simdea/guestlist/ui/main/MainActivity.kt
package pt.simdea.guestlist.ui.main

import android.view.Menu
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import pt.simdea.guestlist.R
import pt.simdea.guestlist.databinding.ActivityMainBinding
import pt.simdea.guestlist.db.model.Item
import pt.simdea.guestlist.ui.base.BaseActivity
import pt.simdea.guestlist.ui.main.list.ListAdapter
import pt.simdea.guestlist.ui.main.list.RowViewHolder
import pt.simdea.guestlist.ui.widgets.MaterialSearchView

class MainActivity : BaseActivity<ActivityMainBinding, MainViewModel>() {

    private val listAdapter: ListAdapter by lazy {
        ListAdapter(::onItemClick)
    }

    override fun layoutToInflate() = R.layout.activity_main

    override fun getViewModelClass() = MainViewModel::class.java

    override fun doOnCreated() {
        setSupportActionBar(dataBinding.toolbar)

        dataBinding.counter.text = viewModel.getCounter().toString()
        dataBinding.searchView.setVoiceSearch(false)
        dataBinding.searchView.setOnQueryTextListener(object : MaterialSearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                addElement(query)
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean = false
        })

        dataBinding.listacompras.setHasFixedSize(true)
        dataBinding.listacompras.setItemViewCacheSize(30)
        dataBinding.listacompras.isNestedScrollingEnabled = false
        dataBinding.listacompras.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        dataBinding.listacompras.adapter = listAdapter
        val itemTouchHelper = ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean = false

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val holder = viewHolder as RowViewHolder

                if (dataBinding.listacompras.adapter!!.itemCount > 0) {
                    val item = holder.item
                    viewModel.deleteItem(item.id)
                }
            }
        })
        itemTouchHelper.attachToRecyclerView(dataBinding.listacompras) //set swipe to recycler view

        viewModel.itemsLiveData?.observe(this, Observer {
            listAdapter.setList(it)
            dataBinding.counter.text = viewModel.getCounter().toString()
        })

    }

    fun addElement(name: String) {
        // save the new Item to the database
        if (name != "") {
            viewModel.addItem(name)
        } else
            Toast.makeText(this, "Item vazio", Toast.LENGTH_LONG).show()
    }

    fun onItemClick(item: Item) {
        viewModel.updateItem(item.id)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)

        val item = menu.findItem(R.id.add)
        dataBinding.searchView.setMenuItem(item)

        return true
    }

    /**
     * {@inheritDoc}
     */
    override fun onBackPressed() {
        if (dataBinding.searchView.isSearchOpen) {
            dataBinding.searchView.closeSearch()
        } else {
            super.onBackPressed()
        }
    }

}
